### PR TITLE
sql/tree: use apd functions to convert decimal to hlc

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Release justification: low risk, high benefit changes to existing functionality

Previously, consumers would often parse a string into an *apd.Decimal,
then while converting to an HLC timestamp this method would re-format
the data in a string, split it using string methods, re-parse it as
integers, before returning the HLC timestamp. Now, native apd methods
are used to reduce the amount of format changes and improve performance.

Release note: None

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
I was slightly worried that the multiplication would be slow, so I also benchmarked the result:

```
# on the previous code
$ go test ./pkg/sql/sem/tree -bench=BenchmarkDecimalToHLC -run=XXX -count 10
I220311 00:55:12.370297 1 (gostd) rand.go:147  [-] 1  random seed: 7209515938303833380
goos: linux
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/sql/sem/tree
cpu: 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
BenchmarkDecimalToHLC-16    	 3669159	       337.9 ns/op
BenchmarkDecimalToHLC-16    	 3469076	       339.4 ns/op
BenchmarkDecimalToHLC-16    	 3373069	       329.1 ns/op
BenchmarkDecimalToHLC-16    	 3634477	       338.9 ns/op
BenchmarkDecimalToHLC-16    	 3649425	       328.9 ns/op
BenchmarkDecimalToHLC-16    	 3368042	       328.5 ns/op
BenchmarkDecimalToHLC-16    	 3599521	       334.4 ns/op
BenchmarkDecimalToHLC-16    	 3666590	       332.6 ns/op
BenchmarkDecimalToHLC-16    	 3231580	       362.4 ns/op
BenchmarkDecimalToHLC-16    	 3533048	       333.5 ns/op
PASS
ok  	github.com/cockroachdb/cockroach/pkg/sql/sem/tree	41.325s

# on the proposed changes
$ go test ./pkg/sql/sem/tree -bench=BenchmarkDecimalToHLC -run=XXX -count 10
I220311 00:53:22.952643 1 (gostd) rand.go:147  [-] 1  random seed: 999453337870328472
goos: linux
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/sql/sem/tree
cpu: 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
BenchmarkDecimalToHLC-16    	 7812388	       186.4 ns/op
BenchmarkDecimalToHLC-16    	 7206136	       165.8 ns/op
BenchmarkDecimalToHLC-16    	 6600723	       156.7 ns/op
BenchmarkDecimalToHLC-16    	 7612414	       157.3 ns/op
BenchmarkDecimalToHLC-16    	 7380686	       170.3 ns/op
BenchmarkDecimalToHLC-16    	 7054263	       169.7 ns/op
BenchmarkDecimalToHLC-16    	 7638564	       163.7 ns/op
BenchmarkDecimalToHLC-16    	 7650508	       171.9 ns/op
BenchmarkDecimalToHLC-16    	 6906928	       164.6 ns/op
BenchmarkDecimalToHLC-16    	 7869111	       156.6 ns/op
PASS
ok  	github.com/cockroachdb/cockroach/pkg/sql/sem/tree	62.397s
$ benchstat /tmp/old.txt /tmp/new.txt 
name             old time/op  new time/op  delta
DecimalToHLC-16   334ns ± 2%   166ns ±12%  -50.16%  (p=0.000 n=9+10)
```

Ended up being a nice speed-up :)

**Comment by stevekuznetsov** [comment]
@mjibson since your name was on the TODO, could you please be a reviewer?

**Comment by stevekuznetsov** [comment]
I understand I need to re-generate some Bazel artifacts, but ...

```
$ ./dev generate bazel
ERROR: please run `dev doctor` to refresh dev status, then try again
$ ./dev doctor
doctor: encountered 1 errors
Could not find STABLE_BUILD_GIT_TYPE in /home/stevekuznetsov/.cache/bazel/_bazel_stevekuznetsov/23c1389ea1eb8b60c96c168fad0d640d/execroot/cockroach/bazel-out/k8-fastbuild/bin/build/bazelutil/test_stamping.txt

This may be because your Bazel is not configured to "stamp" built executables.
Make sure one of the following lines is in the file /home/stevekuznetsov/code/cockroachdb/cockroach/src/github.com/cockroachdb/cockroach/.bazelrc.user:
    build --config=dev
             OR       
    build --config=crosslinux
The former will use your host toolchain, while the latter will use the cross-compiler that we use in CI.
ERROR: please address the errors described above and try again
$ cat /home/stevekuznetsov/code/cockroachdb/cockroach/src/github.com/cockroachdb/cockroach/.bazelrc.user
build --config=dev
```

Bazel has always been a mystery to me so I'm unfortunately stuck here ...

**Comment by jlinder** [comment]
Hi @stevekuznetsov , This is a known error that https://github.com/cockroachdb/cockroach/pull/77605 fixes. Once that PR merges, rebase and you should be unblocked.

**Comment by stevekuznetsov** [comment]
It's not entirely clear how important it is to keep the content of the error messages identical to before, would like to hear how much of this regex-based testing is part of the user-facing API of CRDB and how much is not.

**Comment by otan** [comment]
i would say error messages are flexible here (but minimal changes preferred); i'd ask that you add unit tests to `DecimalToHLC` however.

**Comment by stevekuznetsov** [comment]
@otan OK! I changed up the messages for some of the edge cases as previously they were simply passing along the string parsing errors from the standard library - LMK if you'd rather me try to match them. I was trying to find the existing tests for this one, but I couldn't, so I added test cases for `DecimalToHLC()`.

**Review by rafiss — CHANGES_REQUESTED** [review]
lgtm! i just had minor comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-:-MxuCddX9WYD4_qXbQDw:b-pvle1z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @stevekuznetsov)

___
*[pkg/sql/sem/tree/as_of_test.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-MxuByNSB_c-9xPKtAHT:-MxuByNV4rXncVooYRbD:b-v8c8ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L25)):*
> ```Go
> 
> func BenchmarkDecimalToHLC(b *testing.B) {
> 	r := rand.New(rand.NewSource(timeutil.Now().UnixNano()))
> ```

nit: we prefer to use  rand.NewTestRand(): https://github.com/cockroachdb/cockroach/blob/830283baa3ca349f5be0b42c6a8aef7f155ea18e/pkg/util/randutil/rand.go#L71

it logs the seed so we can reproduce results more easily if needed.
___
*[pkg/sql/sem/tree/as_of_test.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-Mxu5_CR54uUAcPdXL4r:-Mxu5_CWDcqQvVYtseOo:bmdhb8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L104)):*
> ```Go
> 			}
> 			actual, err := DecimalToHLC(dec)
> 			t.Log(err)
> ```

i think we don't need to log the err - if it's unexpected we'll see from the failure message
___
*[pkg/sql/sem/tree/as_of_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-Mxu5hrZ-c2koOaL8Yh2:-Mxu5hrZ-c2koOaL8Yh3:b-ky47fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L107)):*
> ```Go
> 			if err == nil && testCase.err {
> 				t.Fatal("expected an error but got none")
> 			}
> ```

instead of testCase.err being a bool, it could be a string. then this test can use `testutils.IsError` to match the error regex. (empty string means no error is expected)

**Review by stevekuznetsov — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77652)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @stevekuznetsov)

___
*[pkg/sql/sem/tree/as_of_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-Mxu5hrZ-c2koOaL8Yh2:-MxuO1aC6IrWBQdzZZAs:b-doxnxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L107)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of testCase.err being a bool, it could be a string. then this test can use `testutils.IsError` to match the error regex. (empty string means no error is expected)
</blockquote></details>

Do we *want* to do that? I am unfamiliar with the conventions and expectations of this project, but in my past experience in Go if the library wants to expose specific errors in its public API it might do so with:

```go
const ErrLogicalClockTooLarge = errors.New("logical clock too large")
```

So that callers may actually use:

```go
if hlc, err := DecimalToHLC(dec); errors.Is(err, hlc.ErrLogicalClockTooLarge) {
  // do something special in this case
}
```

I was expecting not to have specific errors as part of the API for this function, so I did not validate unexported/internal behavior specifics.

**Review by stevekuznetsov — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77652)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/sem/tree/as_of_test.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-Mxu5_CR54uUAcPdXL4r:-MxuPDpdD5b2NzjH8aCm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L104)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we don't need to log the err - if it's unexpected we'll see from the failure message
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
thanks for your contribution!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-:-MxuPtE638FuqFiHSayF:b3akoa5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @stevekuznetsov)

___
*[pkg/sql/sem/tree/as_of_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77652#-Mxu5hrZ-c2koOaL8Yh2:-MxuPRlt95T0UK1YkAlB:bpijot0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7966670764f14d9672914ff905c4553ef25eda4e/pkg/sql/sem/tree/as_of_test.go#L107)):*
<details><summary><i>Previously, stevekuznetsov (Steve Kuznetsov) wrote…</i></summary><blockquote>

Do we *want* to do that? I am unfamiliar with the conventions and expectations of this project, but in my past experience in Go if the library wants to expose specific errors in its public API it might do so with:

```go
const ErrLogicalClockTooLarge = errors.New("logical clock too large")
```

So that callers may actually use:

```go
if hlc, err := DecimalToHLC(dec); errors.Is(err, hlc.ErrLogicalClockTooLarge) {
  // do something special in this case
}
```

I was expecting not to have specific errors as part of the API for this function, so I did not validate unexported/internal behavior specifics. 
</blockquote></details>

the error text is consumable by end users of the SQL interface, even though the Go error type is not

however, i just realized that the error messages are checked in sql/as_of_test.go so don't think we need to do it here.

**Comment by rafiss** [comment]
i will merge this once the CI passes tests

**Comment by stevekuznetsov** [comment]
@rafiss done! Something about the edit config in this project has made GoLand forget how to `gofmt -s` ... need to look into that.

**Comment by rafiss** [comment]
thanks!

bors r+

**Pull request merged**
